lib/df-detection.c

```diff
@@ -0,0 +1,275 @@
+#include "linux/df-detection.h"
+#include <asm/syscall.h>
+#include <linux/debugfs.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/random.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+#include <linux/sys.h>
+#include <linux/types.h>
+
+void add_address(const void *addr, size_t len, unsigned long caller,
+		 void *kernel_addr)
+{
+	if (!in_task() || current->addresses == NULL || current->pairs == NULL ||
+	    addr > TASK_SIZE)
+		return;
+	char buf[64], buf_caller[64];
+	int size;
+	size = scnprintf(buf, sizeof(buf), "%ps", caller);
+	if (len > MAX_LEN || addr == 0 || strnstr(buf, "copy_from_user_nmi", size))
```

Comment:
Can you check whether we still need copy_from_user_nmi()? Would be nice to remove it.